File: packages/typescript/tests/validators-vectors.test.ts
import fs from 'fs';
import path from 'path';

import { validateRfc } from '../src/validators/rfc';
import { validateCurp } from '../src/validators/curp';
import { validateClabe } from '../src/validators/clabe';
import { validateNss } from '../src/validators/nss';

type ValidationVector = {
  value: string;
  valid: boolean;
};

function loadVectors(fileName: string): ValidationVector[] {
  const filePath = path.resolve(__dirname, `../../shared-data/tests/${fileName}`);
  return JSON.parse(fs.readFileSync(filePath, 'utf-8')) as ValidationVector[];
}

describe('shared validator vectors', () => {
  test.each(loadVectors('rfc_validation.json'))('RFC %s', (vector) => {
    expect(validateRfc(vector.value)).toBe(vector.valid);
  });

  test.each(loadVectors('curp_validation.json'))('CURP %s', (vector) => {
    expect(validateCurp(vector.value)).toBe(vector.valid);
  });

  test.each(loadVectors('clabe_validation.json'))('CLABE %s', (vector) => {
    expect(validateClabe(vector.value)).toBe(vector.valid);
  });

  test.each(loadVectors('nss_validation.json'))('NSS %s', (vector) => {
    expect(validateNss(vector.value)).toBe(vector.valid);
  });
});

File: packages/dart/test/validators_vectors_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:catalogmx/catalogmx.dart';
import 'package:test/test.dart';

List<Map<String, dynamic>> _loadVectors(String name) {
  final file = File('../shared-data/tests/$name');
  final data = jsonDecode(file.readAsStringSync()) as List<dynamic>;
  return data.cast<Map<String, dynamic>>();
}

void main() {
  test('RFC validator shared vectors', () {
    for (final vector in _loadVectors('rfc_validation.json')) {
      expect(validateRFC(vector['value'] as String), vector['valid']);
    }
  });

  test('CURP validator shared vectors', () {
    for (final vector in _loadVectors('curp_validation.json')) {
      expect(validateCURP(vector['value'] as String), vector['valid']);
    }
  });

  test('CLABE validator shared vectors', () {
    for (final vector in _loadVectors('clabe_validation.json')) {
      expect(validateCLABE(vector['value'] as String), vector['valid']);
    }
  });

  test('NSS validator shared vectors', () {
    for (final vector in _loadVectors('nss_validation.json')) {
      expect(validateNSS(vector['value'] as String), vector['valid']);
    }
  });
}
